pytorch_lightning/loops/batch/flexible_optimization_flow.py

```diff
@@ -0,0 +1,161 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+import logging
+from collections import OrderedDict
+from copy import copy
+from typing import Iterator, List, Optional, Tuple
+
+import torch
+
+import pytorch_lightning as pl
+from pytorch_lightning.trainer.progress import OptimizationProgress
+from pytorch_lightning.trainer.supporters import TensorRunningAccum
+from pytorch_lightning.utilities import AttributeDict
+from pytorch_lightning.utilities.batch_loop_common import (
+    check_finite,
+    check_training_step_output,
+    process_training_step_output,
+)
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.model_helpers import is_overridden
+
+log = logging.getLogger(__name__)
+
+
+class FlexibleOptimizationFlow:
+    """
+    The flow for performing a training iteration when ``training_step`` needs access to the
+    dataloader. It is selected when the signature of ``training_step`` contains ``dataloader_iter``:
+
+        def training_step(self, dataloader_iter: Iterator) -> STEP_OUTPUT:
+
+    The ``training_step`` is allowed to fetch multiple batches during one training iteration. The
+    framework provides minimum amount of automation with regards to model optimization (hence the
+    "flexible" in the name). The flexibility allows for ease of experimentation with inter-batch
+    parallelism techniques.
+
+    This flow doesn't support ``automatic_optimization`` and ``tbptt``. An error will be thrown if
+    the ``LightningModule`` or the ``Trainer`` is configured to use these features.
+
+    The ``training_step`` is responsible for reporting whether it has reached the last batch by
+    including an ``is_last`` field in the result dict. Failing to do so will result in an error.
+
+    The ``training_step`` should only optimize the model with one batch for the sake of API and
+    reporting consistency (TODO: consider removing this limitation).
+
+    Args:
+        trainer_ref: a reference to the trainer
+        model_ref: a reference to the lightning module (for config validation purposes only)
+    """
+
+    def __init__(self, trainer_ref: "pl.Trainer", model_ref: "pl.LightningModule") -> None:
+        if is_overridden("on_train_batch_start", model_ref):
+            raise MisconfigurationException(
+                "The model hook `on_train_batch_start` is not compatible with FlexibleOptimizationFlow."
```

Comment:
The users don't know about the `FlexibleOptimizationFlow` class. These errors should say something like

```
not compatible with taking a `dataloader_iter` argument in your `training_step`
```

Reply:
Good call. Will make sure the term doesn't appear in warnings and errors.